MarkBind/markbind
# Add optional mermaid plugin for alternative diagram & chart support

**What is the purpose of this pull request?**




**Overview of changes:**
Closes #1079
Related to #1663 for pie-chart
Related to #984 for diagramming tool

Add an optional plugin that can be enabled to support mermaid-js integration. Done by providing a custom tag and works via [Calling the JavaScript API](https://mermaid-js.github.io/mermaid/#/n00b-gettingStarted?id=_3-calling-the-javascript-api) method. This seems to be the most lightweight approach.

**Anything you'd like to highlight/discuss:**
From my testing, the mermaid syntax does not work in a panel, which seems to be due to how the mermaid library is loaded on the page and how it acts differently when contained in a panel.

Example (doesn't work in a panel, but works without)
```md
<panel>
<mermaid>
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
</mermaid>
</panel>

<mermaid>
graph TD;
    A-->B;
    A-->C;
    B-->D;
    C-->D;
</mermaid>
```

![image](https://user-images.githubusercontent.com/41845017/207212277-f56e1591-bfb6-41e8-9813-a0f8ecae625f.png)


**Testing instructions:**
- check test outputs
- verify updated docs https://deploy-preview-2052--markbind-master.netlify.app/userguide/usingplugins#plugin-mermaid

**Proposed commit message: (wrap lines at 72 characters)**
Add optional mermaid plugin

Existing MarkBind support for graphs is still limited.

Mermaid-js lets users create diagrams and visualizations using text
and code.

Let's add an optional plugin that users can enable if they choose
to create mermaid-supported diagrams.

This feature is added as an optional plugin as some diagrams
supported by mermaid overlaps with PlantUML, which we support
by default.



---

**Checklist:** :ballot_box_with_check:
